qiskit/transpiler/passes/routing/bip_mapping.py

```diff
@@ -0,0 +1,228 @@
+            max_swaps_inbetween_layers (int):
+                Number of swaps allowed in between layers. If None, automatically set.
+                Large value could decrease the probability to build infeasible BIP problem but also
+                could reduce the chance of finding a feasible solution within the ``time_limit``.
+
+        Raises:
+            MissingOptionalLibraryError: if cplex is not installed.
+        """
+        if not HAS_DOCPLEX or not HAS_CPLEX:
+            raise MissingOptionalLibraryError(
+                libname="bip-mapper",
+                name="BIP-based mapping pass",
+                pip_install="pip install 'qiskit-terra[bip-mapper]'",
+            )
+        super().__init__()
+        self.coupling_map = copy.deepcopy(coupling_map)  # save a copy to modify
+        if self.coupling_map is not None:
+            self.coupling_map.make_symmetric()
+        self.objective = objective
+        self.backend_prop = backend_prop
+        self.time_limit = time_limit
+        self.threads = threads
+        self.max_swaps_inbetween_layers = max_swaps_inbetween_layers
+        # ensure the number of virtual and physical qubits are the same
+        self.requires.append(TrivialLayout(self.coupling_map))
+        self.requires.append(FullAncillaAllocation(self.coupling_map))
+        self.requires.append(EnlargeWithAncilla())
+
+    def run(self, dag):
+        """Run the BIPMapping pass on `dag`, assuming the number of virtual qubits (defined in
+        `dag`) and the number of physical qubits (defined in `coupling_map`) are the same.
+
+        Args:
+            dag (DAGCircuit): DAG to map.
+
+        Returns:
+            DAGCircuit: A mapped DAG. If failing to map, returns the original dag.
+
+        Raises:
+            TranspilerError: if the number of virtual and physical qubits are not the same.
+            AssertionError: if the final layout is not valid.
+        """
+        if self.coupling_map is None:
+            return dag
+
+        if len(dag.qubits) > self.coupling_map.size():
+            raise TranspilerError("More virtual qubits exist than physical qubits.")
+
+        if len(dag.qubits) != self.coupling_map.size():
+            raise TranspilerError(
+                "BIPMapping requires the number of virtual and physical qubits are the same."
+            )
+
+        if self.property_set["layout"]:
+            logger.info("BIPMapping ignores any given layout.")
+
+        original_dag = dag
+
+        dummy_steps = math.ceil(math.sqrt(dag.num_qubits()))
+        if self.max_swaps_inbetween_layers is not None:
+            dummy_steps = max(0, self.max_swaps_inbetween_layers - 1)
+
+        model = BIPMappingModel(
+            dag=dag, coupling_map=self.coupling_map, dummy_timesteps=dummy_steps
+        )
+
+        if len(model.su4layers) == 0:
+            logger.info("BIPMapping is skipped due to no 2q-gates.")
+            return original_dag
```

Comment:
I had second thought on this spec. I'm leaning to return a mapped circuit for a circuit with only 1q-gates (i.e. calling `ApplyLayout` and return dag with canonical qregs). If we stick to return original dag, that requires extra handling of the case of no 2q-gates, i.e. checking if a circuit is composed of only 1q-gates or not, and if yes, running some layout pass and `ApplyLayout` instead of calling `BIPMapping`. @1ucian0 What do you think? Anyway, I'll add a commit to change to return a mapped dag here. I'll revert it if others like the original spec.

Reply:
Changed at 76e8c5e

Reply:
> checking if a circuit is composed of only 1q-gates or not, and if yes, running some layout pass and `ApplyLayout`)

This kind of higher level check, I think is better done at passmanager level. With `CheckMap`, a condition is set to run or not the swapper.

Reply:
It makes sense to me. I've reverted the change in c73af80